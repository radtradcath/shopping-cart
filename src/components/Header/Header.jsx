import { Link } from "react-router-dom";
import styles from "./Header.module.css";

function Header() {
  return (
    <header>
      <span>Logo</span>
      <span>Name</span>
      <ul>
        <li>
          <Link to="/">Home</Link>
        </li>
        <li>
          <Link to="/shop">Shop</Link>
        </li>
        <li>
          <Link to="/cart">Cart</Link>
        </li>
      </ul>
    </header>
  );
}

export default Header;
