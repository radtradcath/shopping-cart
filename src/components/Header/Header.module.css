header {
    display: flex;
    justify-content: space-between;
    align-items: center;

    background-color:antiquewhite;
    padding: 1em;
}

header ul {
    display: flex;
    list-style: none;
    gap: 1em;
    padding: 0;
}