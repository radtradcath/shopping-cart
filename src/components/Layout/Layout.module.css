.layout {
    min-height: 100vh;


    display: grid;
    grid-template-rows: 8% 1fr 8%;
}