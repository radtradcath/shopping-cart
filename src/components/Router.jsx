import { createBrowserRouter, RouterProvider } from "react-router-dom";
import { createContext, useState, useEffect } from "react";
import Home from "./Home/Home";
import Shop from "./Shop/Shop";
import Cart from "./Cart/Cart";
import Jewelery from "./Shop/Jewelery";
import Electronics from "./Shop/Electronics";
import Mens from "./Shop/Mens";
import Womens from "./Shop/Women";

export const CartContext = createContext({
  cart: [],
  setCart: () => {},
  addToCart: () => {},
});

const Router = function Router() {
  const [cart, setCart] = useState([]);
  const [items, setItems] = useState([]);

  useEffect(() => {
    const getItems = (async () => {
      const fetchItems = await fetch("https://fakestoreapi.com/products");
      const parsedItems = await fetchItems.json();

      setItems(parsedItems);
    })();
  }, []);

  const addToCart = (e) => {
    const itemId = +e.target.closest(".itemParent").id;
    const foundInCart = cart.some((el) => el.id === itemId);
    const shopItem = items.find((item) => item.id === itemId);

    setCart(
      foundInCart
        ? cart.map((el) =>
            el.id == itemId ? { ...el, count: el.count + 1 } : el
          )
        : [...cart, { ...shopItem, count: 1 }]
    );

    // setCart(
    //   foundInCart
    //     ? cart.map((el) =>
    //         el.id === itemId ? { id: itemId, count: el.count + 1 } : el
    //       )
    //     : [...cart, { id: itemId, count: 1 }]
    // );
    console.log(cart);
  };

  const router = createBrowserRouter([
    {
      path: "/",
      element: <Home />,
    },
    {
      path: "shop",
      element: <Shop />,
      children: [
        { path: "electronics", element: <Electronics /> },
        { path: "jewelery", element: <Jewelery /> },
        { path: "men's clothing", element: <Mens /> },
        { path: "women's clothing", element: <Womens /> },
      ],
    },
    {
      path: "cart",
      element: <Cart />,
    },
  ]);

  return (
    <CartContext.Provider value={{ cart, setCart, addToCart }}>
      <RouterProvider router={router} />
    </CartContext.Provider>
  );
};

export default Router;
