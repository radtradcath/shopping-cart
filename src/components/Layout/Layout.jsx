import Header from "../Header/Header";
import PropTypes from 'prop-types'
import Footer from "../Footer/Footer";
import styles from "./Layout.module.css";
import { createContext, useState } from "react";

function Layout({ children }) {
  return (
    <div className={styles.layout}>
      <Header />
      {children}
      <Footer />
    </div>
  );
}

Layout.propTypes = {
  children: PropTypes.element,
}

export default Layout;
