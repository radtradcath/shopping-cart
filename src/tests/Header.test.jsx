import { render, screen } from "@testing-library/react";
import { expect } from "vitest";
import { cleanup } from "@testing-library/react";
import Layout from "../components/Layout/Layout";
import Header from "../components/Header/Header";
import Router from "../components/Router";
import userEvent from "@testing-library/user-event";

describe("Test header functionalities", () => {
  it("Match snapshot", () => {
    const { container } = render(
      <Router>
        <Layout />
      </Router>
    );
    expect(container).toMatchSnapshot();
  });

  it("Link to shop correctly", async () => {
    const user = userEvent.setup();

    render(
      <Router>
        <Header />
      </Router>
    );

    const shopLink = screen.getByText('Shop');

    await user.click(shopLink);

    expect(screen.getByRole("heading").textContent).toMatch("Shop Page");
  });

  it("Link to cart correctly", async () => {
    const user = userEvent.setup();

    render(
      <Router>
        <Header />
      </Router>
    );

    const cartLink = screen.getByText('Cart');

    await user.click(cartLink);

    expect(screen.getByRole("heading").textContent).toMatch("Cart Page");
  });
});
