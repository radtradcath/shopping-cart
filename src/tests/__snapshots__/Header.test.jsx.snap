// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test header functionalities > Match snapshot 1`] = `
<div>
  <div
    class="_layout_0c8bf9"
  >
    <header>
      <span>
        Logo
      </span>
      <span>
        Name
      </span>
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/shop"
          >
            Shop
          </a>
        </li>
        <li>
          <a
            href="/cart"
          >
            Cart
          </a>
        </li>
      </ul>
    </header>
    <div
      class="_homeContainer_345077"
    >
      <div
        class="_heroContainer_345077"
      >
        <h1
          class="welcome"
        >
          WELCOME TO MY SHOP
        </h1>
      </div>
      <h2
        class="_previewHeading_345077"
      >
        Check the trending
      </h2>
      <div
        class="_previewShop_345077"
      />
    </div>
    <footer>
      <span>
        Made by 
        <a
          href="http://www.github.com/radtradcath"
        >
          RadTradCath
        </a>
      </span>
    </footer>
  </div>
</div>
`;
